runtime/lua/vim/lsp.lua

```diff
@@ -307,6 +307,261 @@ local function create_and_initialize_client(config)
   return client.id, nil
 end
 
+local client_config_list_keys = {
+  filetypes = true,
+  on_attach = true,
+  on_exit = true,
+  on_init = true,
+  root_markers = true,
+}
+
+--- Similar to `vim.tbl_deep_extend('force', ...)`, except that lists specified in
+--- `client_config_list_keys` are extended instead of replaced.
+--- @param ... vim.lsp.Config
+--- @return vim.lsp.Config
+local function merge_configs(...)
+  local ret = {} --- @type table<any,any>
+
+  for i = 1, select('#', ...) do
+    for k, v in
+      pairs(select(i, ...) or {} --[[@as table<string,any>]])
+    do
+      if ret[k] and client_config_list_keys[k] then
+        ret[k] = vim._ensure_list(ret[k])
+        vim.list_extend(ret[k], vim._ensure_list(v))
+      elseif type(ret[k]) == 'table' and type(v) == 'table' then
+        ret[k] = vim.tbl_deep_extend('force', ret[k], v)
+      else
+        ret[k] = v
+      end
+    end
+  end
+
+  return ret
+end
+
+--- @class vim.lsp.Config : vim.lsp.ClientConfig
+---
+--- See `cmd` in [vim.lsp.ClientConfig].
+--- @field cmd? string[]|fun(dispatchers: vim.lsp.rpc.Dispatchers): vim.lsp.rpc.PublicClient
+---
+--- Filetypes the client will attach to, if activated by `vim.lsp.enable()`.
+--- If not provided, then the client will attach to all filetypes.
+--- @field filetypes? string[]
+---
+--- Directory markers (.e.g. '.git/') where the LSP server will base its workspaceFolders,
+--- rootUri, and rootPath on initialization. Unused if `root_dir` is provided.
+--- @field root_markers? string[]
+---
+--- Predicate used to decide if a client should be re-used. Used on all
+--- running clients. The default implementation re-uses a client if name and
+--- root_dir matches.
+--- @field reuse_client? fun(client: vim.lsp.Client, config: vim.lsp.ClientConfig): boolean
+
+--- Update the configuration for an LSP client.
+---
+--- Use name '*' to set default configuration for all clients.
+---
+--- Can also be table-assigned to redefine the configuration for a client.
+---
+--- Examples:
+---
+--- - Add a root marker for all clients:
+---   ```lua
+---     vim.lsp.config('*', {
+---         root_markers = { '.git' },
+---       })
+---       ```
```

Comment:
the indent seems weird, similar for others below

```suggestion
---   ```lua
---   vim.lsp.config('*', {
---     root_markers = { '.git' },
---   })
---   ```
```

Reply:
It's so the code blocks are rendered as:

![image](https://github.com/user-attachments/assets/f9358a43-be03-49ef-9ae7-63cafb39f250)

If anything, it is missing 1 char of indent.

Reply:
hmm. seems like something the docgen should take care of

Reply:
Good thing we have a fully operational `#trim!` soon ;)

Reply:
I'm not sure that'll help since the parser doesn't use queries. It just walks the tree.